libraries/botbuilder-azure/src/cosmosDbPartitionedStorage.ts

```diff
@@ -229,41 +242,76 @@ export class CosmosDbPartitionedStorage implements Storage {
-                            k,
+                            key,
                             this.cosmosDbStorageOptions.keySuffix,
                             this.cosmosDbStorageOptions.compatibilityMode
                         ),
-                        realId: k,
-                        document: changesCopy,
+                        realId: key,
+                        document: change,
                     });
 
-                    const eTag: string = changes[k].eTag;
-                    const ac =
+                    const accessCondition =
                         eTag !== '*' && eTag != null && eTag.length > 0
                             ? { accessCondition: { type: 'IfMatch', condition: eTag } }
                             : undefined;
+
                     try {
-                        await this.container.items.upsert(documentChange, ac);
+                        await this.container.items.upsert(document, accessCondition);
                     } catch (err) {
-                        this.throwInformativeError('Error upserting document', err);
+                        // This check could potentially be performed before even attempting to upsert the item
+                        // so that a request wouldn't be made to Cosmos if it's expected to fail.
+                        // However, performing the check here ensures that this custom exception is only thrown
+                        // if Cosmos returns an error first.
+                        // This way, the nesting limit is not imposed on the Bot Framework side
+                        // and no exception will be thrown if the limit is eventually changed on the Cosmos side.
+                        const nestingError = this.getNestingError(change);
+
+                        this.throwInformativeError(nestingError ?? 'Error upserting document', err);
                     }
                 }
             )
         );
     }
 
+    // Return an informative error message if upsert failed due to deeply nested data
+    private getNestingError(json: object): string | undefined {
+        const checkDepth = (obj: unknown, depth: number, isInDialogState: boolean) => {
+            if (depth > maxDepthAllowed) {
+                const message = `Maximum nesting depth of ${maxDepthAllowed} exceeded.`;
+
+                if (isInDialogState) {
+                    return (
+                        `${message} This is most likely caused by recursive component dialogs. ` +
+                        'Try reworking your dialog code to make sure it does not keep dialogs on the stack ' +
+                        "that it's not using. For example, consider using ReplaceDialogAsync instead of BeginDialogAsync."
+                    );
+                } else {
+                    return `${message} Please check your data for signs of unintended recursion.`;
+                }
+            } else if (obj && typeof (obj) === "object") {
+                depth++;
```

Comment:
I prefer to keep the state mutation in recursive functions purely in the argument list of the recursive invocation - i.e., `checkDepth(..., depth + 1, ...)` rather than manipulating the value directly. It makes it easier to see all the recursive state changes in one place.

Reply:
I was thinking in terms of reducing the amount of times the calculation takes place, but your preference makes sense too